### ISSUE TITLE
Web interface with file input stream gives a JargonException

### ISSUE THREAD

**Issue by jjkoehorst** [DICE-UNC/jargon#390]
I am not sure where to post this as this is probably an error when using a combination of Vaadin and Jargon.

I have created a basic file browser for our irods system to set metadata and get information etc... Now I am adding download functionalities and this requires an input stream.

When downloading the file from irods to our webserver and then make it available through:

```
return Files.newInputStream(Paths.get("." + item.getPath()));
```

Seems to work  perfectly but then the file is also located on the webserver which yields an additional burden on storage.

A direct stream gives a `JargonException in read is converted to IOException for method contract` with an `error code received from iRODS:-345000`.

I am not sure what it means. I tried to use the following with and without rerouting but that returns the same error. is there another type of inputstream that should be used to be able to stream the file like the `Files.newInputStream` stream?

```
// IRODSFile irodsFile = fileFactory.instanceIRODSFile(item.getPath());
IRODSFileInputStream irodsFileInputStream = fileFactory.instanceIRODSFileInputStreamWithRerouting(item.getPath());
```

**Comment by michael-conway** [comment]
can you share the jargon logging with the stack trace?

**Comment by jjkoehorst** [comment]
The stack trace can be found here https://pastebin.com/29dXaTPd

**Comment by trel** [comment]
```
$ ierror 345000
irods error: -345000 BAD_INPUT_DESC_INDEX 
```

**Comment by jjkoehorst** [comment]
Yes I saw it passing by in the stack trace but the documentation is not very clear to what it means?

**Comment by michael-conway** [comment]
Yeah this smells like a multi-threaded access to an agent problem,
wondering if Vaaden is spinning up a new thread somewhere and there's no
real file handle in the new thread.

The link for the stack trace doesn't work so if it could be something
dumped into github as a file that'd be great otherwise we can figure
something else out? If we need to come up with some sort of detachable
thread for these streams we can work on this, it'd be handy.

**Comment by jjkoehorst** [comment]
@michael-conway does this link work better? https://gist.github.com/jjkoehorst/500ab3c8d882fb14eb680fd309eaa773

**Comment by michael-conway** [comment]
yes thanks

**Comment by michael-conway** [comment]
In looking at the Vaaden documentation I am fairly convinced that under the covers the framework is spawning new threads. I think the solution here is to make a separate input/output stream class that will be 'detachable'. The underlying issue is that iRODS agents are not thread safe, so the input stream is segregated in a threadlocal. When the current stream is passed to vaaden the file handle associated with the agent is not present (it's a new agent). I'll take a look at this and that would be handy in a few other scenarios.
